storage/google/cloud/storage/notification.py

```diff
@@ -28,10 +28,12 @@
 NONE_PAYLOAD_FORMAT = 'NONE'
 
 _TOPIC_REF_FMT = '//pubsub.googleapis.com/projects/{}/topics/{}'
-_PROJECT_PATTERN = r'(?P<project>[a-z]+-[a-z]+-\d+)'
+_PROJECT_PATTERN = r'(?P<project>\A[a-z][a-z0-9-]{4,28}[a-z0-9]\Z)'
 _TOPIC_NAME_PATTERN = r'(?P<name>[A-Za-z](\w|[-_.~+%])+)'
 _TOPIC_REF_PATTERN = _TOPIC_REF_FMT.format(
-    _PROJECT_PATTERN, _TOPIC_NAME_PATTERN)
+    _PROJECT_PATTERN.replace(r'\A', '').replace(r'\Z', ''),
```

Comment:
OK @chemelnucfin so I've actually pulled out my editor and looked at this module wholesale. Final answer: just remove all mention of `\A` and `\Z` (they don't need to be replaced with `^` or `$` either).

Reasoning:

- `_PROJECT_PATTERN` is never a full reg-ex, it's a fragment intended to be used with `_TOPIC_REF_PATTERN`
- It "makes" sense to re-use `_TOPIC_REF_FMT` in this way because `_TOPIC_REF_FMT` is needed in `BucketNotification.create`

I have never seen any code that actually uses `\A` and `\Z`, likely because `\A` and `\Z` are only relevant (as alternatives to `^` and `$`) in multiline mode, and when in multiline mode people don't usually think of "beginning" and "end".

Reply:
@dhermes I would suggest again to leave the code as is if there is nothing wrong with it.  In this case, it might be better safe than sorry rather than having "extra" code, especially since someone may accidentally use PROJECT_PATTERN.